File: templates/python-agent/verify.py
"""
ECDSA signature verification for GitHub Copilot inference requests.

GitHub signs every request with an ECDSA P-256 private key. This module
fetches the current public keys from the GitHub API and verifies the
signature attached to each inbound request.
"""

import hashlib
import base64
import requests as _requests

from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.exceptions import InvalidSignature

GITHUB_PUBLIC_KEYS_URL = "https://api.github.com/meta/public_keys/copilot_api"


def _fetch_public_keys() -> list[dict]:
    """
    Fetches GitHub's current ECDSA public keys.

    In production you should cache this response — it rarely changes
    (GitHub rotates keys infrequently).
    """
    resp = _requests.get(
        GITHUB_PUBLIC_KEYS_URL,
        headers={"User-Agent": "copilot-agent-template"},
        timeout=10,
    )
    resp.raise_for_status()
    return resp.json().get("public_keys", [])


def verify_signature(raw_body: bytes, key_id: str, signature_b64: str) -> None:
    """
    Verifies the ECDSA P-256 signature GitHub sends with every Copilot request.

    Args:
        raw_body:       The raw (unparsed) request body bytes.
        key_id:         Value of the X-GitHub-Public-Key-Identifier header.
        signature_b64:  Value of the X-GitHub-Public-Key-Signature header (base64).

    Raises:
        ValueError:        If the key is not found or the signature is invalid.
        InvalidSignature:  Propagated from the cryptography library.
    """
    keys = _fetch_public_keys()
    matching = next((k for k in keys if k.get("key_identifier") == key_id), None)

    if matching is None:
        raise ValueError(f"No GitHub public key found for key_identifier: {key_id}")

    public_key = serialization.load_pem_public_key(matching["key"].encode())

    if not isinstance(public_key, ec.EllipticCurvePublicKey):
        raise ValueError("Expected an EC public key from GitHub")

    signature_bytes = base64.b64decode(signature_b64)

    # GitHub uses IEEE P1363 encoding; convert to DER for the cryptography lib.
    # P-256 signatures are 64 bytes in P1363 (r || s, each 32 bytes).
    if len(signature_bytes) == 64:
        r = int.from_bytes(signature_bytes[:32], "big")
        s = int.from_bytes(signature_bytes[32:], "big")
        from cryptography.hazmat.primitives.asymmetric.utils import encode_dss_signature
        der_signature = encode_dss_signature(r, s)
    else:
        # Already DER-encoded (fallback)
        der_signature = signature_bytes

    public_key.verify(der_signature, raw_body, ec.ECDSA(hashes.SHA256()))
    # verify() raises InvalidSignature if the check fails; nothing to return.

File: templates/nodejs-agent/src/verify.ts
import * as crypto from 'crypto';

const GITHUB_PUBLIC_KEYS_URL =
  'https://api.github.com/meta/public_keys/copilot_api';

interface GitHubPublicKey {
  key_identifier: string;
  key: string;
  is_current: boolean;
}

/**
 * Fetches GitHub's current ECDSA public keys used to sign Copilot requests.
 * In production you should cache this response (it rarely changes).
 */
async function fetchPublicKeys(): Promise<GitHubPublicKey[]> {
  const response = await fetch(GITHUB_PUBLIC_KEYS_URL, {
    headers: { 'User-Agent': 'copilot-agent-template' },
  });
  if (!response.ok) {
    throw new Error(
      `Failed to fetch GitHub public keys: ${response.status} ${response.statusText}`
    );
  }
  const data = (await response.json()) as { public_keys: GitHubPublicKey[] };
  return data.public_keys;
}

/**
 * Verifies the ECDSA P-256 signature that GitHub attaches to every Copilot
 * inference request. Throws an error if verification fails.
 *
 * @param rawBody      - The raw (unparsed) request body as a Buffer or string
 * @param keyId        - The value of the X-GitHub-Public-Key-Identifier header
 * @param signature    - The value of the X-GitHub-Public-Key-Signature header (base64)
 */
export async function verifySignature(
  rawBody: Buffer | string,
  keyId: string,
  signature: string
): Promise<void> {
  const keys = await fetchPublicKeys();
  const matchingKey = keys.find((k) => k.key_identifier === keyId);

  if (!matchingKey) {
    throw new Error(`No GitHub public key found for key_identifier: ${keyId}`);
  }

  const verify = crypto.createVerify('SHA256');
  verify.update(rawBody);

  const isValid = verify.verify(
    { key: matchingKey.key, format: 'pem', type: 'spki', dsaEncoding: 'ieee-p1363' },
    signature,
    'base64'
  );

  if (!isValid) {
    throw new Error('Request signature verification failed');
  }
}
